confiture-web-app/src/pages/audit/AuditCreatePage.vue

```diff
@@ -0,0 +1,286 @@
+onBeforeRouteLeave((to) => {
+  if (!isSubmitting.value && !confirmedLeave.value) {
+    leaveModalDestination.value = to.fullPath;
+    showLeaveModal();
+    return false;
+  }
+});
+
+// Display the native browser confirm modal when leaving site
+function onBeforeUnload(e: BeforeUnloadEvent) {
+  e.preventDefault();
+}
+
+onMounted(() => {
+  window.addEventListener("beforeunload", onBeforeUnload);
+});
+
+onUnmounted(() => {
+  window.removeEventListener("beforeunload", onBeforeUnload);
+});
+
+const accountStore = useAccountStore();
+
+// Steps management
+const currentStep = ref(0);
+const steps = [
+  "Choisissez un type d’audit",
+  "Renseignez l’échantillon des pages à auditer",
+  ...(accountStore.account && accountStore.account.name
+    ? []
+    : ["Indiquez vos coordonnées"])
+];
+const stepHeadingRef = ref<HTMLHeadingElement>();
+
+// Setup audit object
+const audit = ref<CreateAuditRequestData>({
+  auditType: null,
+  procedureName: "",
+  pages: [{ name: "", url: "" }],
+  auditorEmail: accountStore.account?.email ?? "",
+  auditorName: accountStore.account?.name ?? ""
+});
+
+// Default pages per audit type
+const fullDefaultPages = [
+  { name: "Accueil", url: "" },
+  { name: "Contact", url: "" },
+  { name: "Mentions légales", url: "" },
+  { name: "Accessibilité", url: "" },
+  { name: "Plan du site", url: "" },
+  { name: "Aide", url: "" },
+  { name: "Authentification", url: "" }
+];
+
+const fastAndComplementaryDefaultPages = [
+  { name: "Accueil", url: "" },
+  { name: "Contact", url: "" }
+];
+
+// Update default pages except if pages has been changed by user
+const pagesArePristine = ref(true);
+
+const auditType = computed(() => {
+  return audit.value.auditType;
+});
+
+watch(auditType, (newValue) => {
+  if (
+    (newValue === AuditType.FAST || newValue === AuditType.COMPLEMENTARY) &&
+    pagesArePristine.value
+  ) {
+    audit.value.pages = [...fastAndComplementaryDefaultPages];
+  }
+
+  if (newValue === AuditType.FULL && pagesArePristine.value) {
+    audit.value.pages = [...fullDefaultPages];
+  }
+});
+
+// 1. Audit type step
+async function submitAuditTypeStep({
+  auditType,
+  procedureName
+}: {
+  auditType: AuditType;
+  procedureName: string;
+}) {
+  audit.value.auditType = auditType;
+  audit.value.procedureName = procedureName;
+
+  currentStep.value = 1;
+
+  await nextTick();
+  stepHeadingRef.value?.focus();
+}
+
+// 2. Pages step
+async function submitAuditPages(pages: Omit<AuditPage, "id" | "order">[]) {
+  audit.value.pages = pages;
+
+  if (steps.length === 2) {
+    submitAuditSettings();
+  } else {
+    currentStep.value = 2;
+
+    await nextTick();
+    stepHeadingRef.value?.focus();
+  }
+}
+
+// 3. Personal infos and final step
+function submitContactDetailsStep({
+  email,
+  name
+}: {
+  email: string;
+  name: string;
+}) {
+  audit.value.auditorEmail = email;
+  audit.value.auditorName = name;
+
+  if (steps.length === 3) {
+    submitAuditSettings();
+  }
+}
```

Comment:
Tout ça pourrait être la même fonction

Un truc du genre

```typescript
async function submitStep(payload: Partial<CreateAuditRequestData>) {
  audit.value = { ...audit.value, ...payload };

  if (currentStep.value === steps.length) {
    // Send audit creation request when last step is submitted...
    submitAuditSettings()
  } else {
    // ...advance to next step otherwise
    currentStep.value += 1;
    await nextTick();
    stepHeadingRef.value?.focus()
  }
}
```

Reply:
El famoso `Partial` ! J'ai refacto ça du coup :)